File: 965 - Univalued Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isUnivalTree(self, root: TreeNode) -> bool:
        self.val = 114514
        self.unival = True
        self.traverse(root)
        return self.unival
    
    def traverse(self, root):
        if root is None:
            return
        if root.val != self.val:
            if self.val == 114514:
                self.val = root.val
            else:
                self.unival = False
                return
        self.traverse(root.left)
        self.traverse(root.right)

File: 965 - Univalued Binary Tree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    private int val = 1145141919;
    private boolean unival = true;

    public boolean isUnivalTree(TreeNode root) {
        val = 114514;
        unival = true;
        traverse(root);
        return unival;
    }

    private void traverse(TreeNode root) {
        if (root == null) return;
        if (root.val != val) {
            if (val == 114514) val = root.val;
            else {
                unival = false;
                return;
            }
        }
        traverse(root.left);
        traverse(root.right);
    }
}